CAF-2946: Add hash-on-store methods to DataStore

pcrookshpe opened the pull request:
https://jira.autonomy.com/browse/CAF-2946

The surprisingly awkward code in hashStore(InputStream,...) is an attempt to hash the stream while it is being read for storing.

pcrookshpe committed:
CAF-2946: Adding hash-on-store methods to DataStore.

pcrookshpe committed:
CAF-2946: Switching from sha256 to sha1 to match the current hashing performed by BinaryHashWorker.

pcrookshpe commented:
The JIRA ticket https://jira.autonomy.com/browse/CAF-2946 says to extend DataStore but that meant adding default methods to avoid breaking existing implementations of the interface, resulting in the DataStore interface getting more complex than we might have realised when story-pointing the ticket. Let me know if this is acceptable...

kploch commented on worker-api/src/main/java/com/hpe/caf/api/worker/HashStoreResult.java:
```diff
@@ -0,0 +1,63 @@
+        return reference;
+    }
+
+    public void setReference(final String reference)
```
I don't think you need setters on this type. And you can make the reference and hash final keeping this type immutable.

pcrookshpe committed:
CAF-2946: Correcting comments naming SHA variant. Also dropping redundant setters.

pcrookshpe committed:
CAF-2946: Making hash store result type immutable.

dermot-hardy reviewed (changes requested):
Looking at the size of this change, and especially the fact that it has required two extra dependecies to be added to worker-api, which is essentially meant to be an interface-only module, I think we need to move this logic out into another utility module, util-store perhaps.

One utility method would do - we don't really need the byte[] or Path versions.

dermot-hardy reviewed (changes requested):
As discussed, try using a `HashingInputStream` instead.  I would expect it to be much simpler.

pcrookshpe committed:
CAF-2946: Creating a new module util-store for hash-on-store methods.

pcrookshpe committed:
CAF-2946: Adding util-store to worker-framework pom.

dermot-hardy commented on util-store/src/main/java/com/hpe/caf/util/store/StoreUtil.java:
```diff
@@ -0,0 +1,95 @@
+     * @return absolute reference to the stored data, which can be used to retrieve, plus a SHA-1 hash of the stored data
+     * @throws DataStoreException if the data store cannot service the request
+     */
+    public static HashStoreResult hashStore(final DataStore dataStore, final Path dataPath, final String partialReference)
```
This method is reading the file twice.  If it just used the first hashStore() overload that takes a stream it would only be reading the file once.

dermot-hardy commented on worker-store-fs/src/test/java/com/hpe/caf/worker/datastore/fs/FileSystemDataStoreTest.java:
```diff
@@ -17,10 +17,10 @@
-import com.hpe.caf.api.worker.DataStoreException;
-import com.hpe.caf.api.worker.FilePathProvider;
-import com.hpe.caf.api.worker.ReferenceNotFoundException;
+import com.hpe.caf.api.worker.*;
```
You should try to avoid the star import.  It is considered bad practice:
http://checkstyle.sourceforge.net/config_imports.html#AvoidStarImport

pcrookshpe committed:
CAF-2946: Removing inefficient stream usage from hash-on-store methods.
Also, removing wildcarded imports.
Also, corrected and extracted pre-existing duplicated code in FS datastore tests.
